### ISSUE TITLE
"FollowPath" Did not work. 

### ISSUE THREAD

**Issue by maxprzemo** [PokemonGoF/PokemonGo-Bot#5362]
### Expected Behavior
Working "FollowPath"

### Actual Behavior
A week ago everything worked (I have not used a bot by week) . No changes to the settings (I do not see the changes to be introduced in config.json).

### FULL config.json
http://hastebin.com/wuduwiqaci.js

### Output when issue occurred
[2016-09-10 09:49:07] [MainThread] [FollowPath] [INFO] Great job, lot of calories burned! Taking a break now for 0:27:15, will resume at 10:16:22.

The first point of "path_file" It is the end of the loop.
I cleaned folder /PokemonGo-Bot/data , I cleaned files cells-XXXXX.json, last-location-XXXXXX.json, XXX.db, mqtt_client_id

"MoveToFort" works correctly

### Other Information
OS: 
Debian 8.5
Branch: 
master
Git Commit: 
./setup.sh -r  today morning

**Comment by lntn** [comment]
How do you define "not working".

You didn't provide any problem statement in your post.

**Comment by maxprzemo** [comment]
I wrote
The first point of "path_file" It is the end of the loop.
 [INFO] Great job, lot of calories burned! Taking a break now for 0:27:15, will resume at 10:16:22

Bot should visit over a dozen points. And not only the first.
I do not see error in the file .json

[
	{"location": "51.649248, 17.815931"},
	{"location": "51.649298, 17.817352"},
	{"location": "51.650875, 17.817767"},
	{"location": "51.650092, 17.821311"},
	{"location": "51.648132, 17.821406"},
	{"location": "51.647845, 17.819335"},
	{"location": "51.647224, 17.817184"},
	{"location": "51.647391, 17.814938"},
	{"location": "51.645923, 17.814339"},
	{"location": "51.643867, 17.814540"},
	{"location": "51.644918, 17.813323"},
	{"location": "51.647485, 17.811350"},
	{"location": "51.649158, 17.811037"},
	{"location": "51.648709, 17.812522"}
]
A week ago everything worked.
I do not see in the documentation of changes.

**Comment by Anakin5** [comment]
Need more log, giving us only the last line is not going to help a lot. Where are the lines "Walking from..." and every other logs output by the FoolowPath task ?

**Comment by maxprzemo** [comment]
http://hastebin.com/ucefotejef.vbs

**Comment by Anakin5** [comment]
It only looks like you are already at he final location but cannot confirm because you removed your starting position..
Maybe put more loop than only one in your config.

**Comment by maxprzemo** [comment]
"MoveToFort" -> "walker":  "StepWalker" It works
"MoveToFort" -> "walker": "PolylineWalker" It works
"FollowPath" -> "walker": "StepWalker" It works
"FollowPath" -> "walker": "PolylineWalker" Does not want to work

I'm sorry that it did not immediately turned off "walker_limit_output": false,
Now you can see in the log much more
http://hastebin.com/hiwalaredu.coffee


See "distance left:"
Each step is another target.
13 steps to 13 targets with my file path.json
Bot has walked only a few meters
![zrzut ekranu z 2016-09-10 18 53 34](https://cloud.githubusercontent.com/assets/4314537/18412168/980dc5f2-7788-11e6-9d19-0f28f6040e39.png)

**Comment by Anakin5** [comment]
Should be fine now after my PR is merged

**Comment by th3w4y** [comment]
@maxprzemo your output seems correct
We don't handle yet displaying of the real walking distances when using PolylineWalker
What you see there as an increase and decrease in distance is the fact that you took a corner....

>Now you can see in the log much more
http://hastebin.com/hiwalaredu.coffee

**Comment by th3w4y** [comment]
The image with the map you attached does not seam to much anything from the json locations you attached... which street is that pokestop you just visited on?

<img width="331" alt="screen shot 2016-09-11 at 22 59 16" src="https://cloud.githubusercontent.com/assets/10997651/18420459/6a3ef8f8-7873-11e6-8eea-b328bd9d8322.png">

**Comment by Anakin5** [comment]
I fix the issue, his problem comes from the `dist < 1` in FollowPath task.

**Comment by th3w4y** [comment]
@Anakin5 yes you fixed it i confirm!

I was providing an comment regarding the decrease-increase in the displayed distance.

**Comment by th3w4y** [comment]
Wait a second...
http://hastebin.com/hiwalaredu.coffee

The walking speed is strange... 200m/s

Don't want to zoom yet... @Anakin5 was that related with the bug you fixed... is there any possibility that would have manifested with this speed bumps?

```
[2016-09-10 18:52:33] [MainThread] [FollowPath] [INFO] Walking from (51.649009968668885, 17.817190031250046, 135.2562561035156) to (51.645923, 17.814339, 8.0), distance left: (0.40 km) ..
[2016-09-10 18:52:34] [MainThread] [FollowPath] [INFO] Walking from (51.64898998661595, 17.81721003114866, 135.2562561035156) to (51.643867, 17.81454, 8.0), distance left: (0.60 km) ..
[2016-09-10 18:52:35] [MainThread] [FollowPath] [INFO] Walking from (51.64897092415763, 17.81725439513088, 135.4530792236328) to (51.644918, 17.813323, 8.0), distance left: (0.53 km) ..
[2016-09-10 18:52:36] [MainThread] [FollowPath] [INFO] Walking from (51.648959984976194, 17.817280059939826, 135.5178833007812) to (51.647485, 17.81135, 8.0), distance left: (0.44 km) ..
```

**Comment by Anakin5** [comment]
@th3w4y I have still 2 issues to fix in a commit PR. One is related to the fact that we don't always sleep 1 second between step(). The effect is that we can have sudden speed bump because step is called several times per second.
The second issue is that there is a scenario in which we never converge to the final destination.

**Comment by th3w4y** [comment]
close enough to return True you mean?
>The second issue is that there is a scenario in which we never converge to the final destination.

**Comment by th3w4y** [comment]
@solderzzc when do we plan to merge dev into master?

**Comment by Anakin5** [comment]
@th3w4y When we do not converge, it is not because we do not get close enough. I do not have yet the source of the problem but at some point and under certain conditions, the get_next_position enter an infinite loop and repeatedly make us navigate in a very small area without getting closer to the destination, ever.

**Comment by th3w4y** [comment]
@Anakin5 is dev right?

It is in stepWalker

because i don't see the offset_angle formula anymore...
 offset_angle = (1/self.speed)*(precision/1.74)
where did that one disappeared?

**Comment by th3w4y** [comment]
`random_azi = uniform(inverse["azi1"] - 90, inverse["azi1"] + 90)` is much more then precision + epsilon

**Comment by Anakin5** [comment]
It is an angle, precision is a distance

**Comment by th3w4y** [comment]
yes but you are not using the angle based on precision anymore...

**Comment by th3w4y** [comment]
the angle cannot be random... it need to reflect the precision....

**Comment by th3w4y** [comment]
you are calculated the theoretical distance then offset by and angle.... (if that is more then offset_angle it can tick back....)

**Comment by Anakin5** [comment]
The important thing to see in this block is `next_lat, next_lng`

```
random_azi = uniform(line.azi1 - 90, line.azi1 + 90)
random_dist = uniform(0.0, self.precision)
direct = Geodesic.WGS84.Direct(next_lat, next_lng, random_azi, random_dist)
```

Anyway, I think the issue comes from the MoveToFort task. I have a scenario in which the MoveToFort makes us switch between 2 forts forever.

**Comment by th3w4y** [comment]
offset_angle was averring that the angle of azimuth modification is towards the direction of walking....

**Comment by Anakin5** [comment]
I let you read it and draw it on paper. You will see that the angle between origin and the final computed next is always towards the destination.

**Comment by th3w4y** [comment]
![untitled](https://cloud.githubusercontent.com/assets/10997651/18434311/bb910c9a-78ec-11e6-92ec-91dba9d61495.png)

**Comment by Anakin5** [comment]
![image](https://cloud.githubusercontent.com/assets/13438363/18434803/ef1d7280-7921-11e6-968f-d59be60b2e90.png)

**Comment by th3w4y** [comment]
In the destination vicinity....

<img width="514" alt="screen shot 2016-09-12 at 13 52 28" src="https://cloud.githubusercontent.com/assets/10997651/18434877/2f272aa6-78f0-11e6-916f-7f872e5ffde0.png">

**Comment by Anakin5** [comment]
I don't the problem on your graph, Starting from origin, any point in the red area is a valid point and contribute to go toward the destination. The starting point is Origin, it is not 1st next. I highlighted 1st next because we compute two next_point in the get_next_destination function.

**Comment by th3w4y** [comment]
check my update on your graph...
https://github.com/PokemonGoF/PokemonGo-Bot/issues/5362#issuecomment-246324525

**Comment by Anakin5** [comment]
Can you put a legend ? I don't understand what is the new black line and the new black point.

**Comment by th3w4y** [comment]
new black point is the final destination....

**Comment by Anakin5** [comment]
so the red half circle is around it like that:
![image](https://cloud.githubusercontent.com/assets/13438363/18435353/280dc010-7925-11e6-972a-f5fa380c077f.png)

**Comment by th3w4y** [comment]
Ohh i see so is self correcting in next step you mean...

**Comment by Anakin5** [comment]
You are never more that `precision` away from the theoretical path, and you move along the theoretical path. So mathematically, at some point, your are less than `precision` away from your destination.

You can, in extreme random cases, get a bit further from your destination than you were at previous point, but that is giving you more chance to get closer at the next step.

**Comment by th3w4y** [comment]
Yes.. i see i was missing that would self correct on next point...

**Comment by th3w4y** [comment]
it also might not.... settle ever...

follow the next extreme points as being chosen in vicinity or (precision distance) from origin
in this order 1st, 2nd,3rd...

![untitled](https://cloud.githubusercontent.com/assets/10997651/18435967/3c6527b2-78f6-11e6-87a8-e8acc9b36136.png)

**Comment by th3w4y** [comment]
Follow this points 1st, 2nd 3rd...

It might not settle ever...
![untitled](https://cloud.githubusercontent.com/assets/10997651/18436026/960b6880-78f6-11e6-9d50-f24f3582decd.png)

**Comment by Anakin5** [comment]
The center of the red circle is not the origin point (the previous step point). It is the next calculated point along the theoretical path.

**Comment by th3w4y** [comment]
Could still be a problem if the values returned by Geodesic.WGS84.Direct are not within our defined epsilon... (precission)

**Comment by th3w4y** [comment]
Hey I am thinking that if you change that angle to +/- 60 you will converge directly... and will have less problems...

Otherwise you might  have a worker Task that does not receive the proper awaited  "True" and does not do its stuff at destination...

WorkerTask():
1. identifyDestinations_based_on_distance()!!!
2. create_step_walker()
3. step_walker.step()
4. wait _for_true() 
5. do stuff if True... (at destination...)

When such a Task loops... it will select a new destination at part (1.) but never did received True in ( 4.) does would not execute (5.)

**Comment by Anakin5** [comment]
You are supposed to call step() until it is true. You cannot predict the fact that the walker might be a polyline and thus might go away from your destination because that is how the road is.
So even if you choose a destination at 10 meters from you and your speed is 10, it is not a guarantee that the next step will return true.

**Comment by th3w4y** [comment]
I am referring to smaller distances...  around 1m.....

But in principle if +/- 60 degrees should be safer/correct why not use that?

**Comment by Anakin5** [comment]
Can you describe the scenario in which you see it diverging ? Is it a bit more than  precision or a bit less ?

**Comment by th3w4y** [comment]
Or this one if not step_walker.step()


```
        if pokemon['is_vip'] or nearest_fort is None:
            # lock catching(with pokemon_id specified) while moving to vip pokemon or no fort around
            self.bot.capture_locked = pokemon['pokemon_id']
            step_walker = self._move_to(pokemon)
            if not step_walker.step():

                if pokemon['dist'] < Constants.MAX_DISTANCE_POKEMON_IS_REACHABLE:
                    self._encountered(pokemon)
                    self.bot.capture_locked = False  # unlock catch_worker
                    self.inspect(pokemon)
                    return WorkerResult.SUCCESS
                else:
                    return WorkerResult.RUNNING

        else:
            step_walker = self._move_to_pokemon_througt_fort(nearest_fort, pokemon)
            if not step_walker or not step_walker.step():
                return WorkerResult.RUNNING
```

**Comment by Anakin5** [comment]
Which part do you think is broken ? What is the scenario ?

**Comment by th3w4y** [comment]
@Anakin5 I am really sorry bit busy at work cannot write much but... if we don't return the correct status in the current step... next step might be to late for the MoveToMap....

Also I am wondering was the Old StepWalker returning False!!! when the destination was very close... 

Now is returning True when destination is very close...

Logic in MoveToMap might be old.. and not do the proper thing...

Sorry again i cannot continue looking at this... I hope you pursue from this hints...

**Comment by avexus** [comment]
Question, what is the diameter with in the path will the bot spin the poke stop?
I have a path file too. copied from some map that i know of. when i copied those coordinates from google map, and put them into the path file, only half of the stops are spun.
